### PR TITLE
ENH: scipy.sparse.lil: tocsr accelerated

### PR DESCRIPTION
**Pull request by jor-**
This PR accelerates the `tocsr` method of the `lil_matrix` in `scipy.sparse` by about 35%. This also accelerates other methods of the `lil_matrix` which use `tocsr`.

### CODE REVIEW

File: scipy/sparse/lil.py [hunk 1]
```diff
@@ -451,11 +451,18 @@ def tolil(self, copy=False):
 
     def tocsr(self, copy=False):
         lst = [len(x) for x in self.rows]
-        idx_dtype = get_index_dtype(maxval=max(self.shape[1], sum(lst)))
-
-        indptr = np.cumsum([0] + lst, dtype=idx_dtype)
-        indices = np.array([x for y in self.rows for x in y], dtype=idx_dtype)
-        data = np.array([x for y in self.data for x in y], dtype=self.dtype)
+        indptr_dtype = get_index_dtype(maxval=sum(lst))
```
Comment (perimosocordiae) [inline scipy/sparse/lil.py]: As far as I'm aware, most of scipy.sparse assumes that `indptr` and `indices` have the same dtype.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]
Can you run the sparse conversion benchmarks with this change?

File: scipy/sparse/lil.py (hunk 1 above)
Comment (jor-) [reply scipy/sparse/lil.py]: Is this used explicitly somewhere? I have used different dtypes without problems and this could save some memory.

**Comment by jor-** [comment]
I can not run the benchmarks due to a compiler error in scipy/cluster/_hierarchy.c. However I did a simple benchmark by myself:
```
A = scipy.sparse.random(10**4, 10**4, density=0.5, format='lil')
%timeit A.tocsr()
```
results in 
```
9.01 s ± 1.1 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
for the old implementation and
```
5.9 s ± 324 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
for the new one.

File: scipy/sparse/lil.py (hunk 1 above)
Comment (pv) [reply scipy/sparse/lil.py]: Stuff gets upcast on every call to any sparsetools c++ function. There's also the chance that some code somewhere breaks with different dtypes.

**Comment by perimosocordiae** [comment]
I was able to run `python3 runtests.py --bench -s sparse` on a clean master branch without errors. Maybe something is up with you local repo?

File: scipy/sparse/lil.py (hunk 1 above)
Comment (jor-) [reply scipy/sparse/lil.py]: Thank you for the explanation. I have changed it to use the same dtype for both arrays.

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
Thanks for the improvement, @jor- !

**Comment by jor-** [comment]
Thanks for merging.

**Comment by perimosocordiae** [comment]
I ran the benchmarks:
```
python3 runtests.py -t sparse.Conversion \
  --bench-compare=a04161d328bb4adf7b427ef52f00857daa656e1d \
  --bench-compare=7d9022349787cd4e0faef55a30cab22e91e37246 
```

with results:
```
      before           after         ratio
     [a04161d3]       [7d902234]
     <master~3>       <master~2>
+      7.14±0.3ms       18.3±0.1ms     2.56  sparse.Conversion.time_conversion('lil', 'csr')
+     7.31±0.07ms       17.9±0.4ms     2.45  sparse.Conversion.time_conversion('lil', 'csc')
+      7.54±0.3ms       18.3±0.2ms     2.42  sparse.Conversion.time_conversion('lil', 'coo')
+      7.97±0.2ms       18.1±0.2ms     2.27  sparse.Conversion.time_conversion('lil', 'bsr')
+      10.9±0.1ms       21.5±0.3ms     1.98  sparse.Conversion.time_conversion('lil', 'dia')
+      26.3±0.2ms       37.1±0.5ms     1.41  sparse.Conversion.time_conversion('lil', 'dok')
+         241±2μs         253±10μs     1.05  sparse.Conversion.time_conversion('csc', 'coo')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

This particular benchmark is using a "poisson2d" sparse matrix generator function, which produces a 10000x10000 sparse matrix with 49600 stored elements (density ~= 0.0005) with roughly 3 nonzero elements per row. That means this PR's change will be looping 10k times, with two 3-element slice assignments each iteration.

For more random sparsity patterns we do indeed see a speedup with this PR, but it does seem that we have potential regression here. Maybe it's possible to select the appropriate construction method with a heuristic based on density?

**Comment by jor-** [comment]
I have run some additional benchmarks and you are right. For very sparse matrices the old implementation is faster. I have done a corse first screening
```
for i in range(2, 5):
    n = 10**i
    for j in range(11):
        density = 2 * j / n
        A = scipy.sparse.random(n, n, density=density, format='lil', random_state=1)
        time = timeit.timeit('A.tocsr()', globals={'A': A}, number=10) / 10
        print(f'time = {time} for density = {density} and n = {n}')
    print()
```
which resulted in
```
time = 0.001644903399994746 for density = 0.5 and n = 100
time = 0.00019785709999950996 for density = 0.05 and n = 100
time = 0.00012173620000339724 for density = 0.005 and n = 100
time = 0.000133487999994486 for density = 0.0005 and n = 100
time = 0.0008571492000100988 for density = 5e-05 and n = 100
time = 0.0005985439000028237 for density = 5e-06 and n = 100
time = 0.0002099291000035919 for density = 5e-07 and n = 100
time = 0.00011260119999860763 for density = 5e-08 and n = 100
time = 0.000288511499991273 for density = 5e-09 and n = 100
time = 0.000401664499997878 for density = 5e-10 and n = 100

time = 0.06432033930000217 for density = 0.5 and n = 1000
time = 0.007248287300001266 for density = 0.05 and n = 1000
time = 0.0012286973000072976 for density = 0.005 and n = 1000
time = 0.0005635041999994428 for density = 0.0005 and n = 1000
time = 0.0009438137999950413 for density = 5e-05 and n = 1000
time = 0.0004142438999906517 for density = 5e-06 and n = 1000
time = 0.0004403786000011678 for density = 5e-07 and n = 1000
time = 0.0004566987999965022 for density = 5e-08 and n = 1000
time = 0.0004688582000085262 for density = 5e-09 and n = 1000
time = 0.000450516700004755 for density = 5e-10 and n = 1000

time = 7.350531852900008 for density = 0.5 and n = 10000
time = 0.7722896329000036 for density = 0.05 and n = 10000
time = 0.0815837614999964 for density = 0.005 and n = 10000
time = 0.01843994079999902 for density = 0.0005 and n = 10000
time = 0.0034300460000054045 for density = 5e-05 and n = 10000
time = 0.003367760599996927 for density = 5e-06 and n = 10000
time = 0.002489492400002291 for density = 5e-07 and n = 10000
time = 0.002493808799999897 for density = 5e-08 and n = 10000
time = 0.002488352800003213 for density = 5e-09 and n = 10000
time = 0.0025254975000052584 for density = 5e-10 and n = 10000
```
for the old implementation and
```
time = 0.0011412805000077242 for density = 0.5 and n = 100
time = 0.00030763020000676986 for density = 0.05 and n = 100
time = 0.0002503762000060306 for density = 0.005 and n = 100
time = 0.0001997468000013214 for density = 0.0005 and n = 100
time = 0.00040971930000068826 for density = 5e-05 and n = 100
time = 0.00048638209999580796 for density = 5e-06 and n = 100
time = 0.0004102751999994325 for density = 5e-07 and n = 100
time = 0.00035007230000019265 for density = 5e-08 and n = 100
time = 0.00029855010000119365 for density = 5e-09 and n = 100
time = 7.300679999389104e-05 for density = 5e-10 and n = 100

time = 0.03268779119999863 for density = 0.5 and n = 1000
time = 0.002942315400002826 for density = 0.05 and n = 1000
time = 0.00038298009999380154 for density = 0.005 and n = 1000
time = 0.00017930170000681755 for density = 0.0005 and n = 1000
time = 0.0001272976999985076 for density = 5e-05 and n = 1000
time = 0.00012493560000166327 for density = 5e-06 and n = 1000
time = 9.544600000026549e-05 for density = 5e-07 and n = 1000
time = 9.631310000486337e-05 for density = 5e-08 and n = 1000
time = 9.789250000267203e-05 for density = 5e-09 and n = 1000
time = 9.674780000068495e-05 for density = 5e-10 and n = 1000

time = 4.357087663599998 for density = 0.5 and n = 10000
time = 0.373037607699996 for density = 0.05 and n = 10000
time = 0.026649270400002933 for density = 0.005 and n = 10000
time = 0.0021128228000065973 for density = 0.0005 and n = 10000
time = 0.0005444353000029877 for density = 5e-05 and n = 10000
time = 0.00020672089999607124 for density = 5e-06 and n = 10000
time = 0.00017744589999892923 for density = 5e-07 and n = 10000
time = 0.00012488200000007054 for density = 5e-08 and n = 10000
time = 0.00022754210000357488 for density = 5e-09 and n = 10000
time = 9.07887000039409e-05 for density = 5e-10 and n = 10000
```
for the new one. Then I did a finer screening
```
for i in range(2, 5):
    n = 10**i
    for j in range(11):
        density = 10 * j / n
        A = scipy.sparse.random(n, n, density=density, format='lil', random_state=1)
        time = timeit.timeit('A.tocsr()', globals={'A': A}, number=10) / 10
        print(f'time = {time} for density = {density} and n = {n}')
    print()
```
which resulted in
```
time = 0.00016922929999054758 for density = 0.0 and n = 100
time = 0.00028819910003221596 for density = 0.1 and n = 100
time = 0.0010682610000003478 for density = 0.2 and n = 100
time = 0.0006503718999738339 for density = 0.3 and n = 100
time = 0.0006601331000183564 for density = 0.4 and n = 100
time = 0.0018006643999797234 for density = 0.5 and n = 100
time = 0.0009602079000160302 for density = 0.6 and n = 100
time = 0.002110519400002886 for density = 0.7 and n = 100
time = 0.001887765599985869 for density = 0.8 and n = 100
time = 0.0014839080999990983 for density = 0.9 and n = 100
time = 0.0021884500999931333 for density = 1.0 and n = 100

time = 0.0004051115000038408 for density = 0.0 and n = 1000
time = 0.0020549207999920327 for density = 0.01 and n = 1000
time = 0.003440184600003704 for density = 0.02 and n = 1000
time = 0.004930729899979269 for density = 0.03 and n = 1000
time = 0.006820348400015064 for density = 0.04 and n = 1000
time = 0.008192889199972343 for density = 0.05 and n = 1000
time = 0.009588778000033927 for density = 0.06 and n = 1000
time = 0.011695950500006802 for density = 0.07 and n = 1000
time = 0.01292200809998576 for density = 0.08 and n = 1000
time = 0.014300893600011477 for density = 0.09 and n = 1000
time = 0.01678728229999251 for density = 0.1 and n = 1000

time = 0.002443540700005542 for density = 0.0 and n = 10000
time = 0.020600359099989873 for density = 0.001 and n = 10000
time = 0.03572863959998358 for density = 0.002 and n = 10000
time = 0.05424270480002633 for density = 0.003 and n = 10000
time = 0.06736802059999718 for density = 0.004 and n = 10000
time = 0.0858504370999981 for density = 0.005 and n = 10000
time = 0.13122642540001833 for density = 0.006 and n = 10000
time = 0.12917281009999898 for density = 0.007 and n = 10000
time = 0.1581834584000262 for density = 0.008 and n = 10000
time = 0.14190013049997105 for density = 0.009 and n = 10000
time = 0.15214001320000534 for density = 0.01 and n = 10000
```
for the old one and
```
time = 0.0013908662000176263 for density = 0.0 and n = 100
time = 0.0011999756000022899 for density = 0.1 and n = 100
time = 0.0006566542999735248 for density = 0.2 and n = 100
time = 0.0014118695999968621 for density = 0.3 and n = 100
time = 0.000877911799989306 for density = 0.4 and n = 100
time = 0.001648245800015502 for density = 0.5 and n = 100
time = 0.001609994899990852 for density = 0.6 and n = 100
time = 0.0016278238999802852 for density = 0.7 and n = 100
time = 0.0016562741000143433 for density = 0.8 and n = 100
time = 0.0022425096000006305 for density = 0.9 and n = 100
time = 0.002517437100004827 for density = 1.0 and n = 100

time = 0.00211749309996776 for density = 0.0 and n = 1000
time = 0.0036621227000068757 for density = 0.01 and n = 1000
time = 0.004631537700015542 for density = 0.02 and n = 1000
time = 0.00576152120002007 for density = 0.03 and n = 1000
time = 0.006615244400018127 for density = 0.04 and n = 1000
time = 0.007418911700005993 for density = 0.05 and n = 1000
time = 0.008282850000023245 for density = 0.06 and n = 1000
time = 0.009299814099995274 for density = 0.07 and n = 1000
time = 0.010018209899999419 for density = 0.08 and n = 1000
time = 0.01067169009997997 for density = 0.09 and n = 1000
time = 0.03233103320003465 for density = 0.1 and n = 1000

time = 0.017083715799981292 for density = 0.0 and n = 10000
time = 0.029322690700018937 for density = 0.001 and n = 10000
time = 0.039039086899992984 for density = 0.002 and n = 10000
time = 0.04884894479996547 for density = 0.003 and n = 10000
time = 0.05810639599999377 for density = 0.004 and n = 10000
time = 0.06747359780001716 for density = 0.005 and n = 10000
time = 0.07623585530000128 for density = 0.006 and n = 10000
time = 0.0841991484999653 for density = 0.007 and n = 10000
time = 0.09282636279999679 for density = 0.008 and n = 10000
time = 0.11700868559996706 for density = 0.009 and n = 10000
time = 0.11195748539998931 for density = 0.01 and n = 10000
```
for the new one.

The results looks like that the breakdown point at 20 to 30 values per row in average. So maybe 25 values are a good values for switching.

Should I updated the code like this? Do I have to make a new PR?

**Comment by perimosocordiae** [comment]
Yeah, addressing this will need a new PR. If you don't have time to work on it, opening an issue with the findings from your last comment would be nice.

If it's not too much trouble, can you plot the old vs new times vs density?

**Comment by jor-** [comment]
I have made a new PR #10939, but for the plot I do not have time right now.

**Pull request merged**